### PR TITLE
Enhancement 737: Support empty-type columns in QueryBuilder operations

### LINKED ISSUE #737: Add treatment for EmptyType in remaining code
The addition of an `EmptyType` column was merged in PR #646. However, there are many places throughout the code which follow the patterns:
```
if constexpr(is_sequence_type(dt)) {
...
} else {
...
}
```
and
```
if constexpr(is_numeric_type(dt)) {
...
} else {
...
}
```
with the `else` clauses implicitly assuming the type is numeric/sequence respectively. Existing tests pass as they do not use `EmptyType` columns, but it is likely that much functionality would not work with this column type, which we should fix.

### LINKED ISSUE #738: Research if specific handling of EmptyType (and other new types) is needed
**Is your feature request related to a problem? Please describe.**
Up until #646 ArcticDB supported only two types strings and numbers. Some functions assumed that if something is not strings it's a number and vice versa. With the appearance of new types (EmptyType, Decimal, Nullable Bolleans, etc...) we might need to change that and add specific handling for each type rather than
```c++
if(is_sequence_type()) {
// handle strings
} else {
// handle everything else the same way
}
```
The functions which do this are listed here:
```c++
VariantData binary_membership(const ColumnWithStrings& column_with_strings, ValueSet& value_set, Func&& func)
VariantData binary_comparator(const Value& val, const ColumnWithStrings& column_with_strings, Func&& func)
VariantData binary_comparator(const ColumnWithStrings& left, const ColumnWithStrings& right, Func&& func)
VariantData binary_comparator(const ColumnWithStrings& column_with_strings, const Value& val, Func&& func)
SegmentInMemoryImpl::visit_field(Callable &&c)
SegmentInMemoryImpl::operator==(const SegmentInMemoryImpl& left, const SegmentInMemoryImpl& right)
FrameSliceMap::FrameSliceMap(const std::shared_ptr<PipelineContext>& context, bool dynamic_schema)
std::shared_ptr<SegmentInMemoryImpl> filter(const util::BitSet& filter_bitset,
                                                       bool filter_down_stringpool=false,
                                                       bool validate=false) const {
Value(const Value& other)
void SumAggregatorData::aggregate(const std::optional<ColumnWithStrings>& input_column, const std::vector<size_t>& groups, size_t unique_values)
void MinMaxAggregatorData::aggregate(const ColumnWithStrings& input_column)
void MaxOrMinAggregatorData::aggregate(const std::optional<ColumnWithStrings>& input_column, const std::vector<size_t>& groups, size_t unique_values)
VariantData binary_membership(const ColumnWithStrings& column_with_strings, ValueSet& value_set, Func&& func) {
VariantData binary_comparator(const Value& val, const ColumnWithStrings& column_with_strings, Func&& func) {
VariantData binary_comparator(const ColumnWithStrings& left, const ColumnWithStrings& right, Func&& func) {
VariantData binary_comparator(const ColumnWithStrings& column_with_strings, const Value& val, Func&& func) {
inline StreamId stream_id_from_segment(const SegmentInMemory &seg, ssize_t row) {
folly::Future<arcticdb::entity::VariantKey> write_frame_data(const TestDataFrame &data_frame, WriterType &writer) {
```

**Describe the solution you'd like**
Research which of the functions need to change and add handling for the new types.

**Describe alternatives you've considered**

**Additional context**
Relates to: #646 and all other PRs adding new types.

### PR DESCRIPTION
**Pull request by alexowens90**
#### Reference Issues/PRs

Closes #737
Closes #738 

#### What does this implement/fix? How does it work (high level)? Highlight notable design decisions.

Support empty-type columns in all places that used `if_numeric_type/is_sequence_type/is_bool_type` with an `else` statement that made assumptions that all types fell into one of those 3 categories.

This mostly involved fixing `QueryBuilder` operations, and adding tests for them in the C++ layer as there is currently no way to create a column of empty-type from the Python layer.

#### Checklist

<details>
  <summary>
   Checklist for code changes...
  </summary>
 
 - [x] Have you updated the relevant docstrings and documentation?
 - [x] Is this contribution tested against [all ArcticDB's features](../docs/mkdocs/docs/technical/contributing.md)?
 - [x] Do all exceptions introduced raise appropriate [error messages](https://docs.arcticdb.io/error_messages/)?
 - [x] Are API changes highlighted in the PR description?
 - [x] Is the PR labelled as enhancement or bug so it appears in autogenerated release notes?
</details>

### CODE REVIEW

File: cpp/arcticdb/column_store/memory_segment_impl.hpp [hunk 1]
```diff
@@ -295,15 +295,20 @@ class SegmentInMemoryImpl {
                 if constexpr (T::DimensionTag::value == Dimension::Dim0) {
                     if constexpr (is_sequence_type(T::DataTypeTag::data_type)) {
                         // test only for now
-                        throw std::runtime_error("string type not implemented");
+                        internal::raise<ErrorCode::E_ASSERTION_FAILURE>("string type not implemented");
```
[the remark this answers is not in the document]
Comment (alexowens90) [reply cpp/arcticdb/column_store/memory_segment_impl.hpp:298]: There is an equivalent `string_at` method, so it is definitely an error to call this function if the column has a sequence type.
I think you're right though that empty-type columns should also throw here.

File: cpp/arcticdb/column_store/memory_segment_impl.hpp [hunk 2]
```diff
@@ -1021,8 +1030,12 @@ class SegmentInMemoryImpl {
                                 } else {
                                     *output_ptr = value;
                                 }
-                            } else {
+                            } else if constexpr(is_numeric_type(TypeDescriptorTag::DataTypeTag::data_type) || is_bool_type(TypeDescriptorTag::DataTypeTag::data_type)){
```
[the remark this answers is not in the document]
Comment (alexowens90) [reply cpp/arcticdb/column_store/memory_segment_impl.hpp:1033]: Agreed, will add

File: cpp/arcticdb/column_store/test/test_memory_segment.cpp [hunk 3]
```diff
@@ -210,6 +229,20 @@ TEST(MemSegment, CloneAndCompare) {
     ASSERT_EQ(equal, true);
 }
 
+TEST(MemSegment, CloneAndCompareWithEmptyTypeColumn) {
+    SegmentInMemory seg;
+    size_t num_rows = 10;
+    auto int_column = std::make_shared<Column>(generate_int_column(num_rows));
+    seg.add_column(scalar_field(int_column->type().data_type(), "int_column"), int_column);
```
[the remark this answers is not in the document]
Comment (alexowens90) [reply cpp/arcticdb/column_store/test/test_memory_segment.cpp:236]: That a mixture of empty and non-empty column types work together.

File: cpp/arcticdb/processing/operation_dispatch_binary.hpp [hunk 4]
```diff
@@ -222,6 +235,9 @@ VariantData binary_comparator(const Value& val, const ColumnWithStrings& column_
 
 template <typename Func>
 VariantData binary_comparator(const ColumnWithStrings& left, const ColumnWithStrings& right, Func&& func) {
+    if (is_empty_type(left.column_->type().data_type()) || is_empty_type(right.column_->type().data_type())) {
```
[the remark this answers is not in the document]
Comment (alexowens90) [reply cpp/arcticdb/processing/operation_dispatch_binary.hpp:238]: Returning an `EmptyResult` is the correct behavior in this case

[the remark this answers is not in the document]
Comment (vasil-pashov) [reply cpp/arcticdb/processing/operation_dispatch_binary.hpp:238]: Does this mean that two empty columns will be always different. Am I missing something?

[the remark this answers is not in the document]
Comment (alexowens90) [reply cpp/arcticdb/processing/operation_dispatch_binary.hpp:238]: Yes, this is the same behaviour we have with dynamic schema if a row-slice doesn't have the column involved in the comparison.

File: cpp/arcticdb/column_store/memory_segment_impl.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (vasil-pashov) [reply cpp/arcticdb/column_store/memory_segment_impl.hpp:302]: I think the error message is misleading:
```suggestion
                        internal::raise<ErrorCode::E_ASSERTION_FAILURE>("scalar_at not supported with string-type columns");
```
